kafka-connect-s3/src/main/java/io/confluent/connect/s3/TopicPartitionWriter.java

```diff
@@ -366,7 +365,7 @@ public Long getOffsetToCommitAndReset() {
   }
 
   public Long currentStartOffset() {
-    return currentStartOffset;
+    return minStartOffset();
```

Comment:
I am not sure I understand the need for this change. From your description
> Otherwise, if IOException happens during partUpload() but before first call to commit(), Connector doesn't rewind kafka offset back in S3SinkTask.java:239, which leads to a data loss.

If that were to happen, wouldn't `currentStartOffset` not be advanced due to not calling commit? So the rewinding in `S3SinkTask.java:239` should rewind to previous `currentStartOffset` , right? Unless we have a path that would advance the offsets without calling commit, which seems like it would cause data loss as you say, regardless of whether the exception was retriable or not. That seems like it would indicate a bigger problem in the connector. For instance, if the connector were to be restarted in the middle of `partUpload()` , before `commit()` , would that cause data loss?

Reply:
Hi @ncliang,

> If that were to happen, wouldn't `currentStartOffset` not be advanced due to not calling commit? So the rewinding in `S3SinkTask.java:239` should rewind to previous `currentStartOffset` , right?

In original code, `currentStartOffset == null` before first call to `commit()`, so in `S3SinkTask.java:239` no rewinding is happening due to failed null check just above. Then, in `S3SinkTask.java:242` new `TopicPartitionWriter` is created, which start processing from where the previous `TopicPartitionWriter` left off (because no rewinding had taken place). So the records consumed by the previous `TopicPartitionWriter` get lost (because they were neither committed to S3 by old `TopicPartitionWriter` nor re-processed by a new `TopicPartitionWriter`).

> Unless we have a path that would advance the offsets without calling commit, which seems like it would cause data loss as you say, regardless of whether the exception was retriable or not.

No, there is no such path. The problem only appears if lines `S3SinkTask.java:242` and `S3SinkTask.java:244` are executed (replacing existing `TopicPartitionWriter` with a new one). And these lines are located inside of `catch (RetriableException e)` block, so they can execute only if we get an instance of `RetriableException`.

> For instance, if the connector were to be restarted in the middle of `partUpload()` , before `commit()` , would that cause data loss?

No, it wouldn't. On restart, all in-memory state got reset. Connector will start consuming from the latest offsets committed to Kafka. And these offsets will include all "uncommitted to S3" records, because offsets are only committed to Kafka after successful commit to S3 (`TopicPartitionWriter.java:610`).